### DISCUSSION TITLE
How to mock `request` in endpoint from function signrature

### DISCUSSION THREAD

**Discussion by agrepravin** [fastapi/fastapi#7747] [Questions]
### Description

I am using `SessionMiddleware` in my project and inject user info when one login.

In some endpoints I use `request: Request` object to get user info from `request.session["user"]`, during unit testing how do I mock `request` object to return mock user info rather than empty dict?

Please find example code below.

```Python
@app.post("/home")
def random_endpint(request: Request):
    username = request.session["user"]["username"]
```

**Comment by jimcarreer** [comment]
Since the behavior you need to modify is in the SessionMiddleware, ~I would simply monkeypatch the dispatch(...) function on that class to just set your mockdata.   dispatch(...) is what ends up calling your handler (eventually).~  Actually it appears that the session middleware is way more primitive than the BaseHTTPMiddleware.  I would still suggest the approach of creating mock middleware though for this behavior:

```python
import pytest
from fastapi import FastAPI
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
from starlette.testclient import TestClient


@pytest.fixture
def mocked_session_client(request):

    # this is a pytest request fixture, it has to be named this way
    username = request.param

    class MockSessionMiddleware(BaseHTTPMiddleware):

        async def dispatch(self, http_request: Request, call_next: RequestResponseEndpoint) -> Response:
            http_request.scope['session'] = {'user': {'username':  username}}
            return await call_next(http_request)

    ###########################################################################
    # Replace this code with your normal test client / app initialization code
    app = FastAPI()
    @app.get('/test')
    def some_get(http_request: Request) -> dict:
        return {'msg': f'Welcome {http_request.session["user"]["username"]}'}
    client = TestClient(app, raise_server_exceptions=False)
    ###########################################################################

    client.username = username  # Add this so its available int the test.
    app.add_middleware(MockSessionMiddleware)
    yield client


@pytest.mark.parametrize('mocked_session_client', ['username1', 'username2'], indirect=True)
def test_with_session(mocked_session_client: TestClient):
    res = mocked_session_client.get('/test')
    assert res.status_code == 200 and res.json() == {'msg': f'Welcome {mocked_session_client.username}'}
```
So caveat: this is technically "modifying the code under test" (your app).  Some developers will poo-poo this, for your situation its probably fine.  An even better test would be determine the cookie encoding that the session middle ware uses (its likely some kind of signed JSON like a JWT sent as a cookie) and create a fixture that generates that given a session dictionary and send that as a cookie header when making your test client calls.  Thats a bit more involved though so I leave it as an exercise to the reader.


**Also** this middleware is ultimately provided by starlette so I would suggest asking for help on their repo.

**Comment by agrepravin** [comment]
@jimcarreer Thanks a lot

**Comment by tiangolo** [comment]
Thanks @jimcarreer for the help here! :cake: :bowing_man: 

And thanks @agrepravin for reporting back and closing the issue :+1:

**Comment by Defman21** [comment]
For anyone who's looking to solve this by creating a fake session cookie, here's what I came up with:

```python
import json
from base64 import b64encode

from fastapi.testclient import TestClient
from itsdangerous import TimestampSigner

from yourpackage.entrypoint import app


def create_session_cookie(data) -> str:
    signer = TimestampSigner(str(settings.cookie_secret_key))

    return signer.sign(
        b64encode(json.dumps(data).encode('utf-8')),
    ).decode('utf-8')


@pytest.fixture(scope='module')
def client() -> TestClient:
    return TestClient(app)


def test_some_route(client):
    resp = client.get('/some/route', cookies={'session': create_session_cookie({'some': 'state'})})
```

[`SessionMiddleware` uses `itsdangerous` under the hood](https://github.com/encode/starlette/blob/master/starlette/middleware/sessions.py#L24) and here's [the line of code](https://github.com/encode/starlette/blob/master/starlette/middleware/sessions.py#L55) that does the serialization.

**Comment by ihor-pyvovarnyk** [comment]
The simples way to mock sessions would probably be something like this:
```
from starlette.requests import Request

with mock.patch.object(Request, "session", {"foo": "bar"}):  # Either as a context manager or a decorator
    ...
```

It's mocking really on the edge of the session interface and skips all the underlying mechanisms of how sessions work but since the session object itself is dict-like, should probably cover most use cases.
